app/controlplane/internal/service/attestationstate.go

```diff
@@ -0,0 +1,134 @@
+	robotAccount := usercontext.CurrentRobotAccount(ctx)
+	if robotAccount == nil {
+		return nil, errors.NotFound("not found", "robot account not found")
+	}
+
+	initialized, err := s.uc.Initialized(ctx, robotAccount.WorkflowID, req.WorkflowRunId)
+	if err != nil {
+		return nil, handleUseCaseErr("state", err, s.log)
+	}
+
+	return &cpAPI.AttestationStateServiceInitializedResponse{
+		Result: &cpAPI.AttestationStateServiceInitializedResponse_Result{
+			Initialized: initialized,
+		}}, nil
+}
+
+func (s *AttestationStateService) Save(ctx context.Context, req *cpAPI.AttestationStateServiceSaveRequest) (*cpAPI.AttestationStateServiceSaveResponse, error) {
+	robotAccount := usercontext.CurrentRobotAccount(ctx)
+	if robotAccount == nil {
+		return nil, errors.NotFound("not found", "robot account not found")
+	}
+
+	encryptionPassphrase, err := encryptionPassphrase(ctx)
+	if err != nil {
+		return nil, errors.Forbidden("forbidden", "failed to authenticate request")
+	}
+
+	if err := s.uc.Save(ctx, robotAccount.WorkflowID, req.WorkflowRunId, req.AttestationState, encryptionPassphrase); err != nil {
+		return nil, handleUseCaseErr("state", err, s.log)
+	}
+
+	return &cpAPI.AttestationStateServiceSaveResponse{}, nil
+}
+
+func (s *AttestationStateService) Read(ctx context.Context, req *cpAPI.AttestationStateServiceReadRequest) (*cpAPI.AttestationStateServiceReadResponse, error) {
+	robotAccount := usercontext.CurrentRobotAccount(ctx)
+	if robotAccount == nil {
+		return nil, errors.NotFound("not found", "robot account not found")
+	}
+
+	encryptionPassphrase, err := encryptionPassphrase(ctx)
+	if err != nil {
+		return nil, errors.Forbidden("forbidden", "failed to authenticate request")
+	}
+
+	state, err := s.uc.Read(ctx, robotAccount.WorkflowID, req.WorkflowRunId, encryptionPassphrase)
+	if err != nil {
+		return nil, handleUseCaseErr("state", err, s.log)
+	}
+
+	return &cpAPI.AttestationStateServiceReadResponse{
+		Result: &cpAPI.AttestationStateServiceReadResponse_Result{
+			AttestationState: state.State,
+		},
+	}, nil
+}
+
+func (s *AttestationStateService) Reset(ctx context.Context, req *cpAPI.AttestationStateServiceResetRequest) (*cpAPI.AttestationStateServiceResetResponse, error) {
+	robotAccount := usercontext.CurrentRobotAccount(ctx)
+	if robotAccount == nil {
+		return nil, errors.NotFound("not found", "robot account not found")
+	}
+
+	if err := s.uc.Reset(ctx, robotAccount.WorkflowID, req.WorkflowRunId); err != nil {
+		return nil, handleUseCaseErr("state", err, s.log)
+	}
+
+	return &cpAPI.AttestationStateServiceResetResponse{}, nil
+}
+
+// In order to encrypt the state at rest, we will encrypt the state using a passphrase
+// which comes in the request and is used to derive an encryption key.
+// The passphrase that we'll use is the robot-account token that only the workload should have.
```

Comment:
It would be great to use something unique per workload, like some kind of machineguid + token. Should we add a todo here to review it later?

Reply:
Added